File: projecteuler/python/p023.py
# https://projecteuler.net/problem=23

from math import sqrt


def proper_divisor(num):
    yield 1

    largest = int(sqrt(num))
    if largest ** 2 == num:
        yield largest
    else:
        largest += 1

    for div in range(2, largest):
        n, r = divmod(num, div)
        if r == 0:
            yield div
            yield n


def isabundant(num):
    return num < sum(proper_divisor(num))


def isabsum(num, abundants):
    abunds = set(filter(lambda x: x < num, abundants))
    for i in abunds:
        if num - i in abunds:
            return True
    return False


def main():
    limit = 28123

    abundants = [i for i in range(1, limit + 1) if isabundant(i)]

    total = 0
    for num in range(1, limit + 1):
        if not isabsum(num, abundants):
            total += num

    print(total)


if __name__ == '__main__':
    main()

File: projecteuler/lua/p023.lua
local sqrt = math.sqrt
local ceil = math.ceil

table.sum = function(list)
  sigma = 0
  for _, v in ipairs(list) do
    sigma = sigma + v
  end
  return sigma
end

function set(list)
  local _set = {}
  for _, v in ipairs(list) do _set[v] = true end
  return _set
end

function getDivisors(n)
  divisors = {}
  for i = 1, n - 1 do
    if n % i == 0 then
      divisors[#divisors + 1] = i
    end
  end
  return divisors
end

function isAbundant(n)
  return table.sum(getDivisors(n)) > n
end

limit = 28123

abundants = {}
for i = 1, limit do
  if isAbundant(i) then
    abundants[#abundants + 1] = i
  end
end

abundantSum = {}
check = set(abundants)
for j = 1, limit do
  for _, v in ipairs(abundants) do
    if check[j - v] then
      abundantSum[#abundantSum + 1] = j
      break
    end
  end
end

allSums = table.sum(abundantSum)
allNums = (limit * (limit + 1)) * (1 / 2)
print(allNums - allSums)
